src/NuGet.Clients/NuGet.VisualStudio.Common/IDE/VsHierarchyUtility.cs

```diff
@@ -23,21 +23,39 @@ public static class VsHierarchyUtility
         public static string GetProjectPath(IVsHierarchy project)
         {
             ThreadHelper.ThrowIfNotOnUIThread();
-
-            project.GetCanonicalName(VSConstants.VSITEMID_ROOT, out string projectPath);
+            var format = project as IPersistFileFormat;
+            format.GetCurFile(out string projectPath, out uint _);
             return projectPath;
         }
 
-        public static bool IsSupported(IVsHierarchy hierarchy, string projectTypeGuid)
+        public static bool IsSupportedByGuid(IVsHierarchy hierarchy)
         {
             ThreadHelper.ThrowIfNotOnUIThread();
 
-            if (IsProjectCapabilityCompliant(hierarchy))
+            var guids = GetProjectTypeGuids(hierarchy);
+
+            if (guids != null)
             {
+                foreach (var guid in guids) // A project is supported only if all guids are supported.
```

Comment:
This looks scary to me.

When project types like Azure Functions App, or Maui, extend the "base" .NET project, do they add a specific project type?  Are we going to cause a maintenance issue, needing to add to the allow-list every time a new project type extends .NET?

What about non-Microsoft project types like Tizen? .NET Nano Framework? Unity? Wilderness Labs' Meadow?  If they want to support NuGet, are they going to have to ask us to add their project type guid to our allow list?

Or am I misunderstanding how project type guids are used by VS and project systems?

Reply:
These guid checks are not new. 
We already have them today when we do the DTE project walk. 
The only difference is that the way we fetch these guids, the hierarchy API suggests that there can more than 1 guid, while DTE.Kind was always only 1 guid.

Reply:
`GetAggregateProjectTypeGuids` is not the equivalent of `Project.Kind`. `GetAggregateProjectTypeGuids` returns all project types participating in a project, such as the case of project flavors, whereas, `Kind` just returns the core project type.

The equivalent of Kind is `IVsHierarchy.GetGuidProperty(ROOT, VSHPROPID_TypeGuid)`.

Reply:
Thanks @davkean. Fixed. 

@zivkan 

In the end, the checks between new and old code are equivalent: 

New code: 
```cs
public static bool IsNuGetSupported(IVsHierarchy hierarchy)
        {
            ThreadHelper.ThrowIfNotOnUIThread();
            if (IsProjectCapabilityCompliant(hierarchy))
            {
                return true;
            }
            return IsSupportedByGuid(hierarchy) && !HasUnsupportedProjectCapability(hierarchy);
        }

public static bool IsSupportedByGuid(IVsHierarchy hierarchy)
        {
            ThreadHelper.ThrowIfNotOnUIThread();
            ErrorHandler.ThrowOnFailure(hierarchy.GetGuidProperty(VSConstants.VSITEMID_ROOT, (int)__VSHPROPID.VSHPROPID_TypeGuid, out Guid pguid));
            var guid = pguid.ToString("B");
            if (ProjectType.IsUnsupported(guid) || !ProjectType.IsSupported(guid))
            {
                return false;
            }
            return true;
        }
```

Old code: 

Unsupported check is in a different place: 

```cs
public static async Task<bool> IsSupportedAsync(EnvDTE.Project envDTEProject)
        {
            Assumes.Present(envDTEProject);
            await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
            var hierarchy = await envDTEProject.ToVsHierarchyAsync();
            if (VsHierarchyUtility.IsProjectCapabilityCompliant(hierarchy))
            {
                return true;
            }

            return envDTEProject.Kind != null && ProjectType.IsSupported(envDTEProject.Kind) && !VsHierarchyUtility.HasUnsupportedProjectCapability(hierarchy);
        }

public static bool IsExplicitlyUnsupported(EnvDTE.Project envDTEProject)
        {
            ThreadHelper.ThrowIfNotOnUIThread();

            return envDTEProject.Kind == null || ProjectType.IsUnsupported(envDTEProject.Kind);
        }
```

Reply:
Resolving this.